### DISCUSSION TITLE
Working through Windows headaches

### DISCUSSION THREAD

**Discussion by ghost** [lumen-oss/rocks.nvim#506] [Q&A] [answered]
I finally seem to have the luarocks part of the equation working, but I'm still running into issues using Rocks on Windows. Don't know if I have neovim/system configuration issues or I'm finding bugs:

### Setup Details
- Windows 10 no admin privileges
- lua/luarocks installed from the neorocks Scoop bucket
- init.lua copied from [manual install instructions](https://github.com/nvim-neorocks/rocks.nvim/wiki/Installing-rocks.nvim-manually,-without-the-installation-script) with `vim.g.rocks_nvim =  { _log_level = vim.log.levels.TRACE }` added at the top
- rocks.nvim installed manually

### Issues
- When I open neovim or press tab for completion in `:Rocks install` several terminal windows pop up then disappear. The first one appears to be attempting to read from the package manifest. It then says something about download failed and attempts to download. The terminal prints a list of all packages, but nothing persists. The terminal disappears before I can read any messages at the end. The next time I open neovim or press tab for completion, the process repeats

- After installing gruvbox I need to add `vim.opt.runtimepath:append(vim.fs.joinpath(rocks_config.rocks_path, "lib", "luarocks", "rocks-5.1", "gruvbox.nvim", "*"))` in my init file. obviously rather not have to do this for other plugins that need to be available at startup.

- `:Rocks install` works sometimes? `:Rocks install gruvbox.nvim:` worked for me, with the addition of having to add the lib path to gruvbox to rtp.

- Trouble.nvim, downloaded, but the only available command is `Trouble Symbols`.  `Trouble qflist toggle` get an "unkown argument qflist"

- `:Rocks install kanagawa.nvim` did download anything at all

-  In none of these cases does the plugin get added to rocks.toml. 

- `:Rocks prune`  also seems to have inconsistent behavior. `:Rocks prune trouble` hangs with nio keeping the "In progress . . . Pruning" message indefinitely, and `:Rocks prune gruvbox.nvim`  runs for a bit then gives this error:
```
Error executing vim.schedule lua callback: ...ppData/Local/nvim-data/rocks/share/lua/5.1/nio\tasks.lua:100: Async task failed without callback: The coroutine failed with this message:
...Data/Local/nvim-data/rocks/share/lua/5.1/nio\control.lua:293: Vim:E715: Dictionary required
stack traceback:
        [C]: in function 'error'
        ...Data/Local/nvim-data/rocks/share/lua/5.1/nio\control.lua:293: in function 'with'
        ...nvim-data/rocks/share/lua/5.1/rocks\operations\prune.lua:38: in function <...nvim-data/rocks/share/lua/5.1/rocks\operations\prune.lua:37>
stack traceback:
        [C]: in function 'error'
        ...ppData/Local/nvim-data/rocks/share/lua/5.1/nio\tasks.lua:100: in function 'close_task'
        ...ppData/Local/nvim-data/rocks/share/lua/5.1/nio\tasks.lua:122: in function 'cb'
        ...ppData/Local/nvim-data/rocks/share/lua/5.1/nio\tasks.lua:188: in function 'waiter'
        ...Data/Local/nvim-data/rocks/share/lua/5.1/nio\control.lua:116: in function 'wake'
        ...Data/Local/nvim-data/rocks/share/lua/5.1/nio\control.lua:129: in function 'set'
        ...im-data/rocks/share/lua/5.1/rocks\operations\helpers.lua:225: in function 'on_exit'
        ...a/Local/nvim-data/rocks/share/lua/5.1/rocks\luarocks.lua:69: in function 'fn'
        vim/_editor.lua:351: in function <vim/_editor.lua:350>
```

### Logs

The only `:Rocks log` I have after opening Neovim and attempting all of the above is

```
[START][2024-08-15 15:55:45] rocks.nvim logging initiated
ERROR | 2024-08-15 15:55:47 | ...ta/Local/nvim-data/rocks/share/lua/5.1/rocks\adapter.lua:46 | Error creating symlink directory: lua (EPERM: operation not permitted: C:/Users/<user>/AppData/Local/nvim-data/rocks/share/lua/5.1 -> C:/Users/<user>/AppData/Local/nvim-data/rocks/rocks_rtp/lua)
```

**Comment by mrcjkb** [comment]
> After installing gruvbox I need to add vim.opt.runtimepath:append(vim.fs.joinpath(rocks_config.rocks_path, "lib", "luarocks", "rocks-5.1", "gruvbox.nvim", "*")) in my init file. obviously rather not have to do this for other plugins that need to be available at startup.

Right, I forgot that rocks.nvim needs the symlinks to be able to `packadd` the plugins. I don't think there's a way around that.
In principle, creating symlinks is possible on Windows (we have an integration test for it). I don't know why it doesn't work on your system.

**Reply by ghost** [reply]
> In principle, creating symlinks is possible on Windows (we have an integration test for it). I don't know why it doesn't work on your system.

I believe on Windows 10, you have to be an admin user, be granted a specific privilege, or have developer mode turned on, none of which I can do. 

I'm used to running into problems with having low privilege levels. Seems like there's not enough of us understandably for tools developers to spend a lot of time catering to.

**Reply by mrcjkb** [reply]
Argh, why is Windows like this? 
I did some reading and apparently you can create junctions and hard links as a non-admin.
I'll look into that.

**Reply by mrcjkb** [reply]
@fingolfin19 version `2.37.1` of rocks.nvim should create junctions on Windows, instead of NTFS symlinks.
According to my research, this shouldn't require admin privileges.

**Reply by ghost** [reply]
> @fingolfin19 version `2.37.1` of rocks.nvim should create junctions on Windows, instead of NTFS symlinks. According to my research, this shouldn't require admin privileges.

That worked for me.  

Installing Trouble works now (I get all subcommands). 

I realized Kanagawa was failing to install because I needed to add `dev` to `:Rocks install kanagawa`

Kanagawa and Gruvbox both have only <colorscheme>.lua added to the rtp path. There's no `colors/<colorscheme>.lua`, so I can load the colorschemes with their `load()` function, but not `:colorscheme gruvbox`. Is that the intended way for colorschemes to work?

So the remaining rough edges I'm finding:

- Terminal popping up with some process trying to read/download a package manifest every time I open nvim or press tab for autocompletion on `:Rocks install`
-  nui interface doesn't close after installing or updating. It just appears to hang indefinitely with the Rocks spinner going
![image](https://github.com/user-attachments/assets/869c914e-dc46-43a5-95f5-6e1dc3f8f424)
- Maybe relatedly if I've already installed a package and try to install a second package `installing . . .` gets added to the nui interface but nothing happens (no terminals open like when a package is actually being installed). I have to restart to install another plugin. `:Rocks prune` also does not work until I restart, and then I get the same error as in my first post.
- Installing with `:Rocks install` does not add a package to rocks.toml (or does that need to be done manually?)

**Reply by mrcjkb** [reply, marked as answer]
The terminals popping up bug should be fixed in verstion `2.38.1`.

> Kanagawa and Gruvbox both have only .lua added to the rtp path. 

This sounds like a partial/corrupt install (possibly related to what's causing the rocks spinner to hang).
I would suggest you clear everything and try again with a fresh install.

> Maybe relatedly if I've already installed a package and try to install a second package 

Yep, we have a semaphore on install tasks to prevent corruption of the install tree. So if one hangs indefinitely, any subsequent ones will be blocked.

> Installing with :Rocks install does not add a package to rocks.toml 

That means something isn't working properly.
If it still doesn't work after a cleanup/fresh install, I suggest trying some of the things out from [this issue](https://github.com/nvim-neorocks/rocks.nvim/issues/166).

**Reply by ghost** [reply]
> The terminals popping up bug should be fixed in verstion `2.38.1`.
> . . . 

Much better! With a clean install everything I've complained about seems to be fixed now.  Thanks for putting the time in to support us poor souls stuck on Windows!

**Reply by mrcjkb** [reply]
🥳

Much of the credit goes to @MeguminSama.
Many thanks 🙏🙏🙏
